add-custom-dynamic-entity-data-source/src/main/java/com/esri/arcgismaps/sample/addcustomdynamicentitydatasource/components/MapViewModel.kt

```diff
@@ -0,0 +1,133 @@
+/* Copyright 2024 Esri
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.esri.arcgismaps.sample.addcustomdynamicentitydatasource.components
+
+import android.app.Application
+import android.util.Log
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.viewModelScope
+import com.arcgismaps.arcgisservices.LabelingPlacement
+import com.arcgismaps.mapping.ArcGISMap
+import com.arcgismaps.mapping.BasemapStyle
+import com.arcgismaps.mapping.Viewpoint
+import com.arcgismaps.mapping.labeling.LabelDefinition
+import com.arcgismaps.mapping.labeling.SimpleLabelExpression
+import com.arcgismaps.mapping.layers.DynamicEntityLayer
+import com.arcgismaps.mapping.layers.Layer
+import com.arcgismaps.mapping.symbology.TextSymbol
+import com.arcgismaps.mapping.view.SingleTapConfirmedEvent
+import com.arcgismaps.realtime.CustomDynamicEntityDataSource
+import com.arcgismaps.realtime.DynamicEntityObservation
+import com.arcgismaps.toolkit.geoviewcompose.MapViewProxy
+import com.esri.arcgismaps.sample.addcustomdynamicentitydatasource.R
+import kotlinx.coroutines.launch
+import java.io.File
+
+class MapViewModel(application: Application) : AndroidViewModel(application) {
+
+    private val TAG = javaClass.simpleName
+
+    private val provisionPath: String by lazy {
+        application.getExternalFilesDir(null)?.path.toString() + File.separator + application.getString(
+            R.string.app_name)
+    }
+
+    // Create a new custom file source.
+    // This takes the path to the simulation file, field name that will be used as the entity id, and the delay between each observation that is processed.
+    // In this example we are using a json file as our custom data source.
+    // This field value should be a unique identifier for each entity.
+    // Adjusting the value for the delay will change the speed at which the entities and their observations are displayed.
+    private val customSource = CustomEntityFeedProvider(
+        "$provisionPath/AIS_MarineCadastre_SelectedVessels_CustomDataSource.jsonl",
+        "MMSI",
+        10
```

Comment:
For better readability, I suggest to name the parameters:
```suggestion
        fileName = "$provisionPath/AIS_MarineCadastre_SelectedVessels_CustomDataSource.jsonl",
        entityIdField = "MMSI",
        delayDuration = 10
```